logger.go

```diff
@@ -199,35 +222,167 @@ func (old *Log) newChildLog(spanSeed spanSeed, description string, settings LogS
 	return log
 }
 
-func (l *Log) enableFlushTimer() {
-	if l.span.buffered {
-		was := atomic.SwapInt32(&l.shared.FlushActive, 1)
-		if was == 0 {
-			l.shared.FlushTimer.Reset(l.shared.FlushDelay)
+func (log *Log) addMyselfAsDependent() bool {
+	if log == log.request {
+		return false
+	}
+	if log.span.detached {
+		log.request.span.dependentLock.Lock()
+		defer log.request.span.dependentLock.Unlock()
+		log.shared.ActiveDetached[log.span.spanNumber] = log
+		return false
+	}
+	log.parent.span.dependentLock.Lock()
+	defer log.parent.span.dependentLock.Unlock()
+	if log.parent.span.activeDependents == nil {
+		log.parent.span.activeDependents = make(map[int32]*Log)
+	}
+	log.parent.span.activeDependents[log.span.spanNumber] = log
+	return len(log.parent.span.activeDependents) == 1
+}
+
+func (log *Log) hasActivity(startFlusher bool) {
+	was := atomic.SwapInt32(&log.span.knownActive, 1)
+	if was == 0 {
+		if log.addMyselfAsDependent() {
+			log.parent.hasActivity(false)
+		}
+		if startFlusher {
+			wasFlushing := atomic.SwapInt32(&log.shared.FlushActive, 1)
+			if wasFlushing == 0 {
+				log.shared.FlushTimer.Reset(log.shared.FlushDelay)
+			}
 		}
 	}
 }
 
+// Done is used to indicate that a log is complete.  All non-Detach()ed
+// logs that were created from this log are also considered complete.
+//
+// Any logging activity after a Done() causes an error to be logged and may
+// trigger a call to Flush().
+//
+// Done can be synchronous or asynchronous depending on the SynchronousFlush
+// settings.
+func (l *Log) Done() {
+	if l.settings.synchronousFlushWhenDone {
+		l.done(true, true, time.Now())
+	} else {
+		go l.done(true, true, time.Now())
+	}
+}
+
+func (log *Log) recursiveDone(done bool) (count int32) {
+	if done {
+		atomic.StoreInt32(&log.span.knownActive, 0)
+		count = atomic.AddInt32(&log.span.doneCount, 1)
+	}
+	if atomic.SwapInt32(&log.span.knownActive, 0) == 1 {
+		log.span.base.Done(time.Now())
+	}
+	deps := func() []*Log {
+		log.span.dependentLock.Lock()
+		defer log.span.dependentLock.Unlock()
+		deps := make([]*Log, 0, len(log.span.activeDependents))
+		for _, dep := range log.span.activeDependents {
+			deps = append(deps, dep)
+		}
+		return deps
+	}()
+	for _, dep := range deps {
+		_ = dep.recursiveDone(done)
+	}
+	return
+}
+
+func (log *Log) done(explicit bool, doUp bool, now time.Time) {
+	postCount := log.recursiveDone(true)
+	if postCount > 1 && explicit {
```

Comment:
Wouldn't we always hit this condition due to line 278? I believe I figured it out if I see the publicly exposed `Flush()` method, which makes sure that `recursiveDone` is called with `false` parameter.